### PR TITLE
fix: Calculate depreciation_amount accurately

### PR DESCRIPTION
**Pull request by GangaManoj**
## Proposed Changes

- While modifying the _Depreciation Schedule_ of an Asset after its sale, the _Depreciation Amount_ is not calculated correctly if _Depreciation Entries_ have been posted for it already. This PR ensures that the _Depreciation Amount_ calculated will be accurate.

<details>
<summary>Example</summary>

Initial Asset Value = Rs 100,000
Expected Value After Useful Life = Rs 10,000
Number of Depreciation = 3
Frequency of Depreciation = 12

Screenshot 2021-09-20 at 4 50 24 AM

_Before:_ 

On selling the Asset on 31/12/2021, the Depreciation Amount for the second row should have been 30,000, but instead it's being calculated incorrectly as 20,000.

Screenshot 2021-09-20 at 4 52 22 AM

_After:_

Screenshot 2021-09-20 at 4 53 26 AM

</details>

- If the asset is then returned, the Depreciation Schedule should be reset, but the Depreciation Amount for subsequent rows are not being calculated accurately. This PR fixes that as well.

<details>
<summary>Example</summary>

_Before:_ 

On returning the Asset, the Depreciation Amount for the third row should have been 30,000, but instead it's being calculated incorrectly as -30,000.

Screenshot 2021-09-27 at 10 24 42 PM

_After:_

Screenshot 2021-09-20 at 4 50 24 AM

</details>

- If an Asset was set to be sold in the future on the day it was scheduled to be depreciated, once the Asset gets returned, the Depreciation Entry posted on its sale would not be reversed, even if the posting date for the Credit Note is in the future. This PR adds an extra condition to let the Depreciation Entry be cancelled in cases like this.

<details>
<summary>Explanation</summary>

Screenshot 2021-09-20 at 4 50 24 AM

If this Asset is set to be sold on 31-12-2021(which is in the future right now), an additional Depreciation Entry would be made for that date. On returning this Asset, the Depreciation Schedule should be reset and the Depreciation Entry posted on its sale should be reversed, but because the sale was made on the scheduled date for depreciation, the Depreciation Entry does not get reversed. 

With the new condition, the Depreciation Entry made on sale will be reversed if the sale was set to happen in the future(i.e. if the Posting Date for the Sales Invoice is in the future) but gets returned, even if the date of sale = scheduled date for depreciation.

</details>

- Add more tests for Depreciation Schedule.
- Categorise tests into test suites.

### CODE REVIEW

**Comment by blewsky** [comment]
Will that also fix https://github.com/frappe/erpnext/issues/27431 ?

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 1]
```diff
@@ -1041,11 +1040,18 @@ def delete_depreciation_entry_made_after_sale(self, asset):
...
-				if not self.sale_was_made_on_original_schedule_date(asset, schedule, row, posting_date_of_original_invoice):
+				if not self.sale_was_made_on_original_schedule_date(asset, schedule, row, posting_date_of_original_invoice) \
+					or self.sale_happens_in_the_future(posting_date_of_original_invoice):
+
 					reverse_journal_entry = make_reverse_journal_entry(schedule.journal_entry)
 					reverse_journal_entry.posting_date = nowdate()
+					frappe.flags.is_reverse_depr_entry = True
 					reverse_journal_entry.submit()
 
+					asset.flags.ignore_validate_update_after_submit = True
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py:1052]:
```suggestion
					frappe.flags.is_reverse_depr_entry = False
					asset.flags.ignore_validate_update_after_submit = True
```

File: erpnext/assets/doctype/asset/test_asset.py [hunk 2]
```diff
@@ -724,32 +1012,36 @@ def create_asset(**args):
...
+		"asset_owner": args.asset_owner or "Company",
+		"is_existing_asset": args.is_existing_asset or 1
 	})
 
 	if asset.calculate_depreciation:
 		asset.append("finance_books", {
-			"depreciation_method": "Straight Line",
-			"frequency_of_depreciation": 12,
-			"total_number_of_depreciations": 5
+			"depreciation_method": args.depreciation_method or "Straight Line",
```
Comment (nextchamp-saqib) [inline erpnext/assets/doctype/asset/test_asset.py:1034]:
```suggestion
			"finance_book": args.finance_book,
			"depreciation_method": args.depreciation_method or "Straight Line",
```

**Comment by nextchamp-saqib** [comment]
@mergifyio backport version-13-hotfix

**Pull request merged**